### DISCUSSION TITLE
metadata userstring modification with xamarin .dll

### DISCUSSION THREAD

**Discussion by alexisspanidis** [icsharpcode/ILSpy#2963] [Help]
Is it possible to modify the metadata userstring with a xamarin .dll ?

**Comment by dgrunwald** [comment]
I'm not sure what exactly you are asking; but ILSpy in general does not modify assemblies, it is a read-only tool.

**Comment by alexisspanidis** [comment]
> I'm not sure what exactly you are asking; but ILSpy in general does not modify assemblies, it is a read-only tool.

Thanks for your quick reply, actually I'm desperately trying to modify a hardcoded url which ILSpy explorer shows it is stored inside metadata->userstring  . No other decompiler I used so far shows the metadata for that Xamarin .dll except from ILSpy.

**Comment by dgrunwald** [comment]
ILSpy nowadays has support for LZ4-compressed Xamarin .dlls (see #2137 for details). That could be a potential reason why other tools refuse to open your .dll.
Open your .dll in a hex editor. If it starts with ASCII "MZ" it's uncompressed; if it starts with "XALZ" it's compressed. [This comment](https://github.com/icsharpcode/ILSpy/issues/2137#issuecomment-735732995) links to a Python script that can decompress such files.

**Comment by alexisspanidis** [comment]
> ILSpy nowadays has support for LZ4-compressed Xamarin .dlls (see #2137 for details). That could be a potential reason why other tools refuse to open your .dll.
> Open your .dll in a hex editor. If it starts with ASCII "MZ" it's uncompressed; if it starts with "XALZ" it's compressed. [This comment](https://github.com/icsharpcode/ILSpy/issues/2137#issuecomment-735732995) links to a Python script that can decompress such files.

I realized that. Actually I was using dnSpy before to decompile it so I could decompress the Xamarin .dlls by using the lz4 and compress back again after modifying it. I'm trying to save the metadata user strings heap as .il code but the file is empty. Any idea about this?
